File: web/src/engine/player.ts
/**
 * Player state. Python `cli/allin/player.py` 1:1 포팅.
 * 상태 보관 + 덱/손패/실드/회복/턴 훅. 카드 효과 해석은 engine.ts.
 */

import { Boon, Card, ClassName, DamageEvent } from "./types";
import { getDeck } from "./data";
import { Rng, createRng } from "./rng";

export const DEFAULT_HP = 100;
export const DEFAULT_HAND_SIZE = 5;
export const DEFAULT_STARTING_HAND = 5;

export interface BeginTurnInfo {
  heal: number;
  poison: number;
  berserkSelfDmg: number;
}

export interface PlayerOptions {
  name: string;
  className: ClassName;
  boon?: Boon | null;
  seed?: number | null;
}

function effectNum(boon: Boon | null, key: string, fallback = 0): number {
  if (!boon) return fallback;
  const v = boon.effect[key];
  return typeof v === "number" ? v : fallback;
}

export class Player {
  readonly name: string;
  readonly className: ClassName;
  readonly boon: Boon | null;

  // 리소스
  maxHp: number;
  hp: number;
  handSize: number;

  // 덱/손패/묘지
  deck: Card[];
  hand: Card[] = [];
  graveyard: Card[] = [];

  // 턴 내 일시 효과
  shield = 0;

  // 다음-1회 효과
  nextAccBonus = 0;
  nextCritBonus = 0;
  guaranteeNextCrit = false;
  dodgeNextPercent = 0;
  nextAttackMissChance = 0;

  // 지속형
  rageStacks = 0;
  poisonTurns = 0;
  poisonDamage = 0;
  silencedCards: string[] = [];
  betCapOverride: number | null = null;
  betCapOverrideTurns = 0;

  // 특수 모드
  berserkTurns = 0;
  berserkAccBonus = 0;
  berserkDamageBonus = 0;

  incomingDamageMult = 1.0;
  incomingDamageMultTurns = 0;

  // 누적 통계
  totalBet = 0;
  totalDamageTaken = 0;
  totalDamageDealt = 0;
  critCount = 0;
  missCount = 0;

  // 추적
  lastCard: Card | null = null;
  missedLastTurn = false;
  private _missedThisTurn = false;
  sigUsedIds = new Set<string>();
  sigUsedThisTurn = false;

  extraCardPerTurn: number;
  lastPeek: Card | null = null; // 워든 패시브 노출

  private rng: Rng;

  constructor(options: PlayerOptions) {
    this.name = options.name;
    this.className = options.className;
    this.boon = options.boon ?? null;
    this.rng = createRng(options.seed ?? null);

    // HP
    this.maxHp = DEFAULT_HP + effectNum(this.boon, "hp_bonus");
    this.hp = this.maxHp;

    // 손패
    this.handSize = effectNum(this.boon, "hand_size", DEFAULT_HAND_SIZE);

    // 덱 셔플
    this.deck = getDeck(this.className);
    this.rng.shuffle(this.deck);

    // BN09
    this.extraCardPerTurn = effectNum(this.boon, "extra_card_per_turn");

    // 시작 손패
    const starting =
      DEFAULT_STARTING_HAND + effectNum(this.boon, "starting_hand_bonus");
    this.draw(starting);
  }

  isAlive(): boolean {
    return this.hp > 0;
  }

  // ---- 덱/손패 -----------------------------------------------------------
  draw(n = 1): Card[] {
    const drawn: Card[] = [];
    for (let i = 0; i < n; i++) {
      if (this.deck.length === 0) {
        if (this.graveyard.length === 0) break;
        this.deck = this.graveyard;
        this.graveyard = [];
        this.rng.shuffle(this.deck);
      }
      const card = this.deck.pop()!;
      this.hand.push(card);
      drawn.push(card);
    }
    return drawn;
  }

  fillHand(): Card[] {
    const need = Math.max(0, this.handSize - this.hand.length);
    return this.draw(need);
  }

  discard(card: Card): void {
    const idx = this.hand.indexOf(card);
    if (idx >= 0) this.hand.splice(idx, 1);
    this.graveyard.push(card);
  }

  redrawHand(): Card[] {
    const count = this.hand.length;
    this.graveyard.push(...this.hand);
    this.hand = [];
    return this.draw(count);
  }

  // ---- 피해/회복 ---------------------------------------------------------
  takeDamage(amount: number, opts?: { ignoreShield?: boolean }): DamageEvent {
    const evt: DamageEvent = {
      requested: amount,
      absorbedByShield: 0,
      reducedByBoon: 0,
      dealt: 0,
      dodged: false,
    };
    if (amount <= 0) return evt;
    const ignoreShield = opts?.ignoreShield ?? false;
    let dealt = amount;

    if (!ignoreShield && this.shield > 0) {
      const absorbed = Math.min(this.shield, dealt);
      evt.absorbedByShield = absorbed;
      dealt -= absorbed;
      this.shield -= absorbed; // 🐛 fix: 방어막이 흡수한 만큼 차감
    }

    const dr = effectNum(this.boon, "damage_reduction");
    if (dr > 0 && dealt > 0) {
      const minDmg = effectNum(this.boon, "min_damage", 1);
      const newDealt = Math.max(minDmg, dealt - dr);
      evt.reducedByBoon = dealt - newDealt;
      dealt = newDealt;
    }

    evt.dealt = dealt;
    this.hp -= dealt;
    this.totalDamageTaken += dealt;
    return evt;
  }

  heal(amount: number): number {
    if (amount <= 0 || !this.isAlive()) return 0;
    const before = this.hp;
    this.hp = Math.min(this.maxHp, this.hp + amount);
    return this.hp - before;
  }

  // ---- 턴 훅 --------------------------------------------------------------
  beginTurn(): BeginTurnInfo {
    const info: BeginTurnInfo = { heal: 0, poison: 0, berserkSelfDmg: 0 };

    // 실드 리셋 (상대 턴까지 유지되다가 내 턴 시작 시 0)
    this.shield = 0;

    // BN03 회복의 가호
    const hpt = effectNum(this.boon, "heal_per_turn");
    if (hpt > 0) {
      info.heal += this.heal(hpt);
    }

    // B15 베르세르크 자해
    if (this.berserkTurns > 0) {
      const selfDmg = 5;
      const before = this.hp;
      this.hp = Math.max(0, this.hp - selfDmg);
      info.berserkSelfDmg = before - this.hp;
      this.berserkTurns -= 1;
      if (this.berserkTurns === 0) {
        this.berserkAccBonus = 0;
        this.berserkDamageBonus = 0;
      }
    }

    // 독
    if (this.poisonTurns > 0) {
      const evt = this.takeDamage(this.poisonDamage, { ignoreShield: true });
      info.poison = evt.dealt;
      this.poisonTurns -= 1;
      if (this.poisonTurns === 0) this.poisonDamage = 0;
    }

    this._missedThisTurn = false;
    this.sigUsedThisTurn = false;

    return info;
  }

  endTurn(): void {
    this.missedLastTurn = this._missedThisTurn;

    if (this.betCapOverrideTurns > 0) {
      this.betCapOverrideTurns -= 1;
      if (this.betCapOverrideTurns === 0) this.betCapOverride = null;
    }

    if (this.incomingDamageMultTurns > 0) {
      this.incomingDamageMultTurns -= 1;
      if (this.incomingDamageMultTurns === 0) this.incomingDamageMult = 1.0;
    }

    this.silencedCards = [];
  }

  // ---- 유틸 ---------------------------------------------------------------
  recordMiss(): void {
    this.missCount += 1;
    this._missedThisTurn = true;
  }

  recordHit(damage: number, critical = false): void {
    this.totalDamageDealt += damage;
    if (critical) this.critCount += 1;
  }

  spendBet(amount: number): number {
    if (amount <= 0) return 0;
    this.hp -= amount;
    this.totalBet += amount;

    let refunded = 0;
    const ratio = effectNum(this.boon, "bet_refund_ratio");
    if (ratio > 0) {
      refunded = Math.floor(amount * ratio);
      this.heal(refunded);
    }
    return amount - refunded;
  }

  maxCardsPerTurn(): number {
    return 2 + this.extraCardPerTurn;
  }
}

File: cli/allin/player.py
"""Player state container. 턴 중/간 상태, 덱/손패/묘지, 상태효과, 누적 통계를 관리.

이 모듈은 '상태 보관 + 기본 조작'에 집중. 카드 효과 해석은 engine.py (Task 4+) 가 담당.
하지만 턴 시작/종료 훅(회복의 가호, 독, 베르세르크, shield 리셋 등)은 여기서 처리해야
엔진 코드가 과하게 복잡해지지 않으므로 함께 구현한다.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import Optional

from .boons import Boon
from .cards import Card, get_deck

DEFAULT_HP = 100
DEFAULT_HAND_SIZE = 5
DEFAULT_STARTING_HAND = 5


@dataclass
class DamageEvent:
    """take_damage 결과를 설명용으로 노출. UI/로그에서 사용."""

    requested: int  # 원래 데미지 값
    absorbed_by_shield: int = 0
    reduced_by_boon: int = 0
    dealt: int = 0  # 실제 HP 차감량
    dodged: bool = False


class Player:
    def __init__(
        self,
        name: str,
        class_name: str,
        boon: Optional[Boon] = None,
        *,
        seed: Optional[int] = None,
    ) -> None:
        self.name = name
        self.class_name = class_name
        self.boon = boon
        self._rng = random.Random(seed)

        # ---- 리소스 ----
        hp_bonus = boon.effect.get("hp_bonus", 0) if boon else 0
        self.max_hp: int = DEFAULT_HP + hp_bonus
        self.hp: int = self.max_hp

        # ---- 덱/손패/묘지 ----
        self.hand_size: int = (
            boon.effect.get("hand_size", DEFAULT_HAND_SIZE) if boon else DEFAULT_HAND_SIZE
        )
        self.deck: list[Card] = get_deck(class_name)
        self._rng.shuffle(self.deck)
        self.hand: list[Card] = []
        self.graveyard: list[Card] = []

        # ---- 턴 내 일시 효과 (턴 종료 시 리셋) ----
        self.shield: int = 0                # B7/B9 "이번 턴 받는 데미지 -X"

        # ---- 다음 N회에만 적용되는 효과 ----
        self.next_acc_bonus: int = 0         # B6 광전사의 함성
        self.next_crit_bonus: int = 0        # G11 미래 보기
        self.guarantee_next_crit: bool = False  # G14 마크된 운명
        self.dodge_next_percent: int = 0     # G9/W12 "다음 받는 공격 X% 회피"
        self.next_attack_miss_chance: int = 0  # G12 블러프 → 내 다음 공격이 X% 빗나감 강제

        # ---- 지속형 상태 ----
        self.rage_stacks: int = 0            # B11 분노 스택 (다음 공격에 소모)
        self.poison_turns: int = 0
        self.poison_damage: int = 0
        self.silenced_cards: list[str] = []  # W3 약점 포착 (카드 id 1턴 봉인)
        self.bet_cap_override: Optional[int] = None   # W4 결박의 사슬: 다음 턴 베팅 상한
        self.bet_cap_override_turns: int = 0

        # ---- 직업/시그니처 특수 모드 ----
        self.berserk_turns: int = 0          # B15 베르세르크
        self.berserk_acc_bonus: int = 0
        self.berserk_damage_bonus: int = 0

        self.incoming_damage_mult: float = 1.0  # W14 무효 선언 (내가 받는 데미지 × mult)
        self.incoming_damage_mult_turns: int = 0

        # ---- 게임 전체 누적 통계 ----
        self.total_bet: int = 0
        self.total_damage_taken: int = 0
        self.total_damage_dealt: int = 0
        self.crit_count: int = 0
        self.miss_count: int = 0

        # ---- 기록/추적 ----
        self.last_card: Optional[Card] = None
        self.missed_last_turn: bool = False    # 가장 최근 턴 동안 miss 낸 적 있음
        self._missed_this_turn: bool = False
        self.hit_last_turn: bool = False       # 가장 최근 턴 동안 적중시킨 적 있음 (W2 조건)
        self._hit_this_turn: bool = False
        self.sig_used_ids: set[str] = set()
        self.sig_used_this_turn: bool = False

        # ---- 부운 기반 초기 설정 ----
        self.extra_card_per_turn: int = (
            boon.effect.get("extra_card_per_turn", 0) if boon else 0
        )

        # ---- 시작 손패 드로우 ----
        starting = DEFAULT_STARTING_HAND + (
            boon.effect.get("starting_hand_bonus", 0) if boon else 0
        )
        self.draw(starting)

    # ------------------------------------------------------------------ 기본 상태
    def is_alive(self) -> bool:
        return self.hp > 0

    # ------------------------------------------------------------------ 덱/손패
    def draw(self, n: int = 1) -> list[Card]:
        """덱에서 n장 뽑아 손패에 추가. 덱이 비면 묘지 셔플해서 복구."""
        drawn: list[Card] = []
        for _ in range(n):
            if not self.deck:
                if not self.graveyard:
                    break
                self.deck = self.graveyard
                self.graveyard = []
                self._rng.shuffle(self.deck)
            card = self.deck.pop()
            self.hand.append(card)
            drawn.append(card)
        return drawn

    def fill_hand(self) -> list[Card]:
        """손패를 hand_size 까지 보충."""
        need = max(0, self.hand_size - len(self.hand))
        return self.draw(need)

    def discard(self, card: Card) -> None:
        if card in self.hand:
            self.hand.remove(card)
        self.graveyard.append(card)

    def redraw_hand(self) -> list[Card]:
        """G10 손패 바꾸기: 전부 버리고 같은 수만큼 드로우."""
        count = len(self.hand)
        self.graveyard.extend(self.hand)
        self.hand = []
        return self.draw(count)

    # ------------------------------------------------------------------ 피해/회복
    def take_damage(self, amount: int, *, ignore_shield: bool = False) -> DamageEvent:
        """피해 적용. shield 감쇠 → 부운 감쇠 → HP 차감 순."""
        evt = DamageEvent(requested=amount)
        if amount <= 0:
            return evt

        dealt = amount

        if not ignore_shield and self.shield > 0:
            absorbed = min(self.shield, dealt)
            evt.absorbed_by_shield = absorbed
            dealt -= absorbed

        # BN02 견고한 의지: 받는 모든 데미지 -2 (최소 1)
        if self.boon and self.boon.effect.get("damage_reduction", 0) > 0 and dealt > 0:
            reduction = self.boon.effect["damage_reduction"]
            min_dmg = self.boon.effect.get("min_damage", 1)
            new_dealt = max(min_dmg, dealt - reduction)
            evt.reduced_by_boon = dealt - new_dealt
            dealt = new_dealt

        evt.dealt = dealt
        self.hp -= dealt
        self.total_damage_taken += dealt
        return evt

    def heal(self, amount: int) -> int:
        """HP 회복, max_hp 상한. 실제 회복량 반환."""
        if amount <= 0 or not self.is_alive():
            return 0
        before = self.hp
        self.hp = min(self.max_hp, self.hp + amount)
        return self.hp - before

    # ------------------------------------------------------------------ 턴 훅
    def begin_turn(self) -> dict:
        """턴 시작 시 자동 효과 적용. 엔진이 호출.

        반환값: {'heal': int, 'poison': int, 'berserk_self_dmg': int} — UI 로깅용.
        """
        info: dict[str, int] = {"heal": 0, "poison": 0, "berserk_self_dmg": 0}

        # 실드는 "이번 턴 받는 데미지 -X" 의미로 쓰이되,
        # 사용 턴 이후 상대 턴까지 유지되어야 방어 카드로서 역할을 함.
        # 따라서 리셋은 사용자 본인이 새 턴을 시작할 때 수행.
        self.shield = 0

        # BN03 회복의 가호: 매 턴 시작 +5
        if self.boon and self.boon.effect.get("heal_per_turn", 0) > 0:
            info["heal"] += self.heal(self.boon.effect["heal_per_turn"])

        # B15 베르세르크: 매 턴 자해 5
        if self.berserk_turns > 0:
            self_dmg = 5
            before = self.hp
            self.hp = max(0, self.hp - self_dmg)
            info["berserk_self_dmg"] = before - self.hp
            self.berserk_turns -= 1
            if self.berserk_turns == 0:
                self.berserk_acc_bonus = 0
                self.berserk_damage_bonus = 0

        # 독 데미지 (방어 무시)
        if self.poison_turns > 0:
            evt = self.take_damage(self.poison_damage, ignore_shield=True)
            info["poison"] = evt.dealt
            self.poison_turns -= 1
            if self.poison_turns == 0:
                self.poison_damage = 0

        # 턴 내 추적자 리셋
        self._missed_this_turn = False
        self._hit_this_turn = False
        self.sig_used_this_turn = False

        return info

    def end_turn(self) -> None:
        """턴 종료 시 정리. 지속 효과 카운터 감소 + 턴 플래그 이월.

        참고: 실드(`shield`) 는 상대 턴에도 방어해야 하므로 begin_turn 에서 리셋.
        """
        # 이번 턴 miss/hit 냈는지 플래그를 '직전 턴' 상태로 이월
        self.missed_last_turn = self._missed_this_turn
        self.hit_last_turn = self._hit_this_turn

        # 결박의 사슬 베팅 상한: 다음 턴 1회만 유효 → 한 턴 지나면 해제
        if self.bet_cap_override_turns > 0:
            self.bet_cap_override_turns -= 1
            if self.bet_cap_override_turns == 0:
                self.bet_cap_override = None

        # 무효 선언: 내가 받는 데미지 감쇠 턴 카운트 감소
        if self.incoming_damage_mult_turns > 0:
            self.incoming_damage_mult_turns -= 1
            if self.incoming_damage_mult_turns == 0:
                self.incoming_damage_mult = 1.0

        # 침묵 해제 (1턴)
        self.silenced_cards = []

    # ------------------------------------------------------------------ 유틸
    def record_miss(self) -> None:
        self.miss_count += 1
        self._missed_this_turn = True

    def record_hit(self, damage_dealt: int, critical: bool = False) -> None:
        self.total_damage_dealt += damage_dealt
        self._hit_this_turn = True
        if critical:
            self.crit_count += 1

    def spend_bet(self, amount: int) -> int:
        """베팅 HP 차감 + 피의 환원 부운 적용. 실제 순지출(차감-환원) 반환."""
        if amount <= 0:
            return 0
        # HP 1 남기기 룰: caller 가 강제할 것. 여기선 단순 차감.
        self.hp -= amount
        self.total_bet += amount

        # BN08 피의 환원: 베팅한 HP의 50% 즉시 회복
        refunded = 0
        if self.boon and self.boon.effect.get("bet_refund_ratio", 0) > 0:
            ratio = self.boon.effect["bet_refund_ratio"]
            refunded = int(amount * ratio)
            self.heal(refunded)

        return amount - refunded

    def max_cards_per_turn(self) -> int:
        return 2 + self.extra_card_per_turn

    # ------------------------------------------------------------------ 디버그
    def __repr__(self) -> str:
        return (
            f"Player(name={self.name!r}, class={self.class_name}, "
            f"hp={self.hp}/{self.max_hp}, hand={len(self.hand)}, "
            f"deck={len(self.deck)}, grave={len(self.graveyard)})"
        )
